src/lib/optimizer/strategy/logical_reduction_rule.cpp

```diff
@@ -0,0 +1,142 @@
+
+bool LogicalReductionRule::apply_to(const std::shared_ptr<AbstractLQPNode>& node) const {
+  Assert(node->type == LQPNodeType::Root, "LogicalReductionRule needs root to hold onto");
+
+  /**
+   * Step 1:
+   *    - Reduce Expressions in the LQP
+   *    - Collect PredicateNodes that can be split up into multiple ones into `predicate_nodes_to_flat_conjunctions`
+   */
+  auto predicate_nodes_to_flat_conjunctions =
+      std::vector<std::pair<std::shared_ptr<PredicateNode>, std::vector<std::shared_ptr<AbstractExpression>>>>{};
+
+  visit_lqp(node, [&](const auto& sub_node) {
+    // We mostly aim at PredicateNodes, since these are the nodes that primarily contain logical expressions.
+    // TODO(anybody) once we have a generic way to rewrite expressions in an arbitrary node, use it here
+
+    if (const auto predicate_node = std::dynamic_pointer_cast<PredicateNode>(sub_node)) {
+      const auto new_predicate = reduce_distributivity(predicate_node->predicate);
+      const auto flat_conjunction = flatten_logical_expressions(new_predicate, LogicalOperator::And);
+
+      if (flat_conjunction.size() > 1) {
+        predicate_nodes_to_flat_conjunctions.emplace_back(predicate_node, flat_conjunction);
+      }
+    } else if (const auto projection_node = std::dynamic_pointer_cast<ProjectionNode>(sub_node)) {
+      for (auto& expression : projection_node->expressions) {
+        expression = reduce_distributivity(expression);
+      }
+    }
+
+    return LQPVisitation::VisitInputs;
+  });
+
+  /**
+   * Step 2:
+   *    - Split up qualifying PredicateNodes into multiple consecutive PredicateNodes. We have to do this in a
+   *      second pass, because manipulating the LQP within `visit_lqp()`, while theoretically possible, is prone to
+   *      bugs.
+   */
+  for (const auto& [predicate_node, flat_conjunction] : predicate_nodes_to_flat_conjunctions) {
+    for (const auto& predicate_expression : flat_conjunction) {
+      lqp_insert_node(predicate_node, LQPInputSide::Left, PredicateNode::make(predicate_expression));
+    }
+    lqp_remove_node(predicate_node);
+  }
+
+  return false;
+}
+
+std::shared_ptr<AbstractExpression> LogicalReductionRule::reduce_distributivity(
+    const std::shared_ptr<AbstractExpression>& input_expression) {
+  // Step 1: `(a AND b AND c) OR (a AND d AND b AND e)` --> `[(a AND b AND c), (a AND d AND b AND e)]`
+  const auto flat_disjunction = flatten_logical_expressions(input_expression, LogicalOperator::Or);
+
+  // Step 2: `[(a AND b AND c), (a AND d AND b AND e)]` --> `[[a, b, c], [a, d, b, e]]`
+  auto flat_disjunction_and_conjunction = std::vector<std::vector<std::shared_ptr<AbstractExpression>>>{};
+  flat_disjunction_and_conjunction.reserve(flat_disjunction.size());
+
+  for (const auto& expression : flat_disjunction) {
+    flat_disjunction_and_conjunction.emplace_back(flatten_logical_expressions(expression, LogicalOperator::And));
+  }
+
+  // Step 3: Identify common_conjunctions: [a, b]
```

Comment:
`std::set_intersection`?

Reply:
Sadly only possible for orderable objects. Expressions, as of now, are not orderable.
(and using the orderable property of the shared_ptrs won't work)